File: src/android/KochavaTrackerPlugin.java
package com.kochava.tracker;

//
//  KochavaTracker (Cordova)
//
//  Copyright (c) 2013 - 2021 Kochava, Inc. All rights reserved.
//

import android.content.Context;
import android.util.Log;

import com.kochava.core.json.internal.JsonArray;
import com.kochava.core.json.internal.JsonArrayApi;
import com.kochava.core.json.internal.JsonObjectApi;
import com.kochava.core.util.internal.ObjectUtil;
import com.kochava.core.util.internal.TextUtil;
import com.kochava.tracker.Tracker;
import com.kochava.tracker.attribution.InstallAttributionApi;
import com.kochava.tracker.attribution.RetrievedInstallAttributionListener;
import com.kochava.tracker.deeplinks.DeeplinkApi;
import com.kochava.tracker.deeplinks.ProcessedDeeplinkListener;
import com.kochava.tracker.engagement.Engagement;
import com.kochava.tracker.events.Event;
import com.kochava.tracker.events.EventApi;
import com.kochava.tracker.events.Events;
import com.kochava.tracker.log.LogLevel;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;

/**
 * Kochava Tracker Plugin
 */
public final class KochavaTrackerPlugin extends CordovaPlugin {
    private static final String LOGTAG = "Kochava/Tracker";
    private Context context;

    /**
     * Initialize the Plugin.
     * @param cordova Cordova interface interface.
     * @param webView Cordova webview interface.
     */
    @Override
    public final void initialize(final CordovaInterface cordova, final CordovaWebView webView) {
        super.initialize(cordova, webView);
        if(context == null && cordova != null && cordova.getActivity() != null) {
            context = cordova.getActivity().getApplicationContext();
        }
    }

    /**
     * Performs an action within the sdk as defined by the action parameter.
     *
     * @param action to perform.
     * @param argsRaw is a list of arguments to pass to the specific action.
     * @param callbackContext to pass to results of the action.
     * @return True if completed and false if unable to complete.
     */
    @Override
    public final boolean execute(final String action, final JSONArray argsRaw, final CallbackContext callbackContext) {
        if(TextUtil.isNullOrBlank(action) || argsRaw == null) {
            return false;
        }
        final JsonArrayApi args = JsonArray.buildWithJSONArray(argsRaw);

        // void executeAdvancedInstruction(string name, string value)
        if (action.equals("executeAdvancedInstruction") && args.length() == 2) {
            final String name = args.getString(0, "");
            final String value = args.getString(1, "");
            Tracker.getInstance().executeAdvancedInstruction(name, value);
            return true;
        }

        // void setLogLevel(LogLevel logLevel)
        if (action.equals("setLogLevel") && args.length() == 1) {
            final LogLevel logLevel = LogLevel.fromString(args.getString(0, ""));
            Tracker.getInstance().setLogLevel(logLevel);
            return true;
        }

        // void setSleep(bool sleep)
        if (action.equals("setSleep") && args.length() == 1) {
            final boolean sleep = args.getBoolean(0, false);
            Tracker.getInstance().setSleep(sleep);
            return true;
        }

        // void setAppLimitAdTracking(bool appLimitAdTracking)
        if (action.equals("setAppLimitAdTracking") && args.length() == 1) {
            final boolean appLimitAdTracking = args.getBoolean(0, false);
            Tracker.getInstance().setAppLimitAdTracking(appLimitAdTracking);
            return true;
        }

        // void registerCustomDeviceIdentifier(string name, string value)
        if (action.equals("registerCustomDeviceIdentifier") && args.length() == 2) {
            final String name = args.getString(0, "");
            final String value = args.getString(1, "");
            Tracker.getInstance().registerCustomDeviceIdentifier(name, value);
            return true;
        }

        // void registerIdentityLink(string name, string value)
        if (action.equals("registerIdentityLink") && args.length() == 2) {
            final String name = args.getString(0, "");
            final String value = args.getString(1, "");
            Tracker.getInstance().registerIdentityLink(name, value);
            return true;
        }

        // void enableAndroidInstantApps(string instantAppGuid)
        if (action.equals("enableAndroidInstantApps") && args.length() == 1) {
            final String instantAppGuid = args.getString(0, "");
            Tracker.getInstance().enableInstantApps(instantAppGuid);
            return true;
        }

        // void enableIosAppClips(string identifier)
        if (action.equals("enableIosAppClips") && args.length() == 1) {
            Log.w(LOGTAG, "enableIosAppClips API is not available on this platform.");
            return true;
        }

        // void enableIosAtt()
        if (action.equals("enableIosAtt") && args.length() == 1) {
            Log.w(LOGTAG, "enableIosAtt API is not available on this platform.");
            return true;
        }

        // void setIosAttAuthorizationWaitTime(double waitTime)
        if (action.equals("setIosAttAuthorizationWaitTime") && args.length() == 1) {
            Log.w(LOGTAG, "setIosAttAuthorizationWaitTime API is not available on this platform.");
            return true;
        }

        // void setIosAttAuthorizationAutoRequest(bool autoRequest)
        if (action.equals("setIosAttAuthorizationAutoRequest") && args.length() == 1) {
            Log.w(LOGTAG, "setIosAttAuthorizationAutoRequest API is not available on this platform.");
            return true;
        }

        // void registerPrivacyProfile(string name, string[] keys)
        if(action.equals("registerPrivacyProfile") && args.length() == 2) {
            final String name = args.getString(0, "");
            final String[] keys = ObjectUtil.jsonArrayToStringArray(args.getJsonArray(1, true));
            Tracker.getInstance().registerPrivacyProfile(name, keys);
            return true;
        }

        // void setPrivacyProfileEnabled(string name, bool enabled)
        if(action.equals("setPrivacyProfileEnabled") && args.length() == 2) {
            final String name = args.getString(0, "");
            final boolean enabled = args.getBoolean(1, false);
            Tracker.getInstance().setPrivacyProfileEnabled(name, enabled);
            return true;
        }

        // bool getStarted()
        if(action.equals("getStarted") && args.length() == 0 && callbackContext != null) {
            callbackContext.success(Boolean.toString(Tracker.getInstance().isStarted()));
            return true;
        }

        // void start(string androidAppGuid, string iosAppGuid, string partnerName)
        if (action.equals("start") && args.length() == 3) {
            final String androidAppGuid = args.getString(0, "");
            // arg 1 is iosAppGuid
            final String partnerName = args.getString(2, "");
            if(!TextUtil.isNullOrBlank(androidAppGuid)) {
                Tracker.getInstance().startWithAppGuid(context, androidAppGuid);
            } else if(!TextUtil.isNullOrBlank(partnerName)) {
                Tracker.getInstance().startWithPartnerName(context, partnerName);
            } else {
                Log.e(LOGTAG, "No App Guid or Partner Name was registered.");
            }
            return true;
        }

        // void shutdown(bool deleteData)
        if (action.equals("shutdown") && args.length() == 1) {
            final boolean deleteData = args.getBoolean(0, false);
            Tracker.getInstance().shutdown(context, deleteData);
            return true;
        }

        // string getDeviceId()
        if(action.equals("getDeviceId") && args.length() == 0 && callbackContext != null) {
            callbackContext.success(Tracker.getInstance().getDeviceId());
            return true;
        }

        // InstallAttribution getInstallAttribution()
        if(action.equals("getInstallAttribution") && args.length() == 0 && callbackContext != null) {
            callbackContext.success(Tracker.getInstance().getInstallAttribution().toJson());
            return true;
        }

        // void retrieveInstallAttribution(Callback<InstallAttribution> callback)
        if(action.equals("retrieveInstallAttribution") && args.length() == 0 && callbackContext != null) {
            Tracker.getInstance().retrieveInstallAttribution(new RetrievedInstallAttributionListener() {
                @Override
                public final void onRetrievedInstallAttribution(final InstallAttributionApi installAttribution) {
                    callbackContext.success(installAttribution.toJson());
                }
            });
            return true;
        }

        // void processDeeplink(string path, Callback<Deeplink> callback)
        if(action.equals("processDeeplink") && args.length() == 1 && callbackContext != null) {
            final String path = args.getString(0, "");
            Tracker.getInstance().processDeeplink(path, new ProcessedDeeplinkListener() {
                @Override
                public final void onProcessedDeeplink(final DeeplinkApi deeplink) {
                    callbackContext.success(deeplink.toJson());
                }
            });
            return true;
        }

        // void processDeeplinkWithOverrideTimeout(string path, double timeout, Callback<Deeplink> callback)
        if(action.equals("processDeeplinkWithOverrideTimeout") && args.length() == 2 && callbackContext != null) {
            final String path = args.getString(0, "");
            final double timeout = args.getDouble(1, 10.0);
            Tracker.getInstance().processDeeplink(path, timeout, new ProcessedDeeplinkListener() {
                @Override
                public final void onProcessedDeeplink(final DeeplinkApi deeplink) {
                    callbackContext.success(deeplink.toJson());
                }
            });
            return true;
        }

        // void registerPushToken(string token)
        if (action.equals("registerPushToken") && args.length() == 1) {
            final String token = args.getString(0, "");
            Engagement.getInstance().registerPushToken(token);
            return true;
        }

        // void setPushEnabled(bool enabled)
        if (action.equals("setPushEnabled") && args.length() == 1) {
            final boolean enabled = args.getBoolean(0, false);
            Engagement.getInstance().setPushEnabled(enabled);
            return true;
        }

        // void sendEvent(string name)
        if (action.equals("sendEvent") && args.length() == 1) {
            final String name = args.getString(0, "");
            Events.getInstance().send(name);
            return true;
        }

        // void sendEventWithString(string name, string data)
        if (action.equals("sendEventWithString") && args.length() == 2) {
            final String name = args.getString(0, "");
            final String data = args.getString(1, "");
            Events.getInstance().sendWithString(name, data);
            return true;
        }

        // void sendEventWithDictionary(string name, object data)
        if (action.equals("sendEventWithDictionary") && args.length() == 2) {
            final String name = args.getString(0, "");
            final JsonObjectApi data = args.getJsonObject(1, true);
            Events.getInstance().sendWithDictionary(name, data.toJSONObject());
            return true;
        }

        // void sendEventWithEvent(Event event)
        if (action.equals("sendEventWithEvent") && args.length() == 1) {
            final JsonObjectApi eventInfo = args.getJsonObject(0, true);
            final String eventName = eventInfo.getString("name", "");
            final JsonObjectApi eventData = eventInfo.getJsonObject("data", true);
            final String androidGooglePlayReceiptData = eventInfo.getString("androidGooglePlayReceiptData", "");
            final String androidGooglePlayReceiptSignature = eventInfo.getString("androidGooglePlayReceiptSignature", "");
            final EventApi event = Event.buildWithEventName(eventName);
            event.mergeCustomDictionary(eventData.toJSONObject());
            if(!TextUtil.isNullOrBlank(androidGooglePlayReceiptData) && !TextUtil.isNullOrBlank(androidGooglePlayReceiptSignature)) {
                event.setGooglePlayReceipt(androidGooglePlayReceiptData, androidGooglePlayReceiptSignature);
            }
            event.send();
            return true;
        }

        Log.i(LOGTAG, "Invalid Action");
        return false;
    }
}

File: src/ios/KochavaTrackerPlugin.m
//
//  KochavaTracker (Cordova)
//
//  Copyright (c) 2013 - 2021 Kochava, Inc. All rights reserved.
//

#pragma mark - Import

#import "KochavaTrackerPlugin.h"

#pragma mark - Util

// Interface for the kochavaTrackerUtil
@interface KochavaTrackerUtil : NSObject

@end

// Common utility functions used by all of the wrappers.
// Any changes to the methods in here must be propagated to the other wrappers.
@implementation KochavaTrackerUtil

// Log a message to the console.
+ (void)log:(nonnull NSString *)message {
    NSLog(@"Kochava/Tracker: %@", message);
}

// Attempts to read an NSDictionary and returns nil if not one.
+ (nullable NSDictionary *)readNSDictionary:(nullable id)valueId {
    return [[NSDictionary class] performSelector:@selector(kva_fromObject:) withObject:valueId];
}

// Attempts to read an NSArray and returns nil if not one.
+ (nullable NSArray *)readNSArray:(nullable id)valueId {
    return [[NSArray class] performSelector:@selector(kva_fromObject:) withObject:valueId];
}

// Attempts to read an NSNumber and returns nil if not one.
+ (nullable NSNumber *)readNSNumber:(nullable id)valueId {
    return [[NSNumber class] performSelector:@selector(kva_fromObject:) withObject:valueId];
}

// Attempts to read an NSString and returns nil if not one.
+ (nullable NSString *)readNSString:(nullable id)valueId {
    return [NSString kva_fromObject:valueId];
}

// Converts an NSNumber to a double with fallback to a default value.
+ (double)convertNumberToDouble:(nullable NSNumber *)number defaultValue:(double)defaultValue {
    if(number != nil) {
        return [number doubleValue];
    }
    return defaultValue;
}

// Converts an NSNumber to a bool with fallback to a default value.
+ (BOOL)convertNumberToBool:(nullable NSNumber *)number defaultValue:(BOOL)defaultValue {
    if(number != nil) {
        return [number boolValue];
    }
    return defaultValue;
}

// Converts the deeplink result into an NSDictionary.
+ (nonnull NSDictionary *)convertDeeplinkToDictionary:(nonnull KVADeeplink *)deeplink {
    NSObject *object = [deeplink kva_asForContextObjectWithContext:KVAContext.host];
    return [object isKindOfClass:NSDictionary.class] ? (NSDictionary *)object : @{};
}

// Converts the install attribution result into an NSDictionary.
+ (nonnull NSDictionary *)convertInstallAttributionToDictionary:(nonnull KVAAttributionResult *)installAttribution {
    if (KVATracker.shared.startedBool) {
        NSObject *object = [installAttribution kva_asForContextObjectWithContext:KVAContext.host];
        return [object isKindOfClass:NSDictionary.class] ? (NSDictionary *)object : @{};
    } else {
        return @{
                @"retrieved": @(NO),
                @"raw": @{},
                @"attributed": @(NO),
                @"firstInstall": @(NO),
        };
    }
}

// Serialize an NSDictionary into a json serialized NSString.
+ (nullable NSString *)serializeJsonObject:(nullable NSDictionary *)dictionary {
    return [NSString kva_stringFromJSONObject:dictionary prettyPrintBool:NO];
}

// Parse a json serialized NSString into an NSArray.
+ (nullable NSArray *)parseJsonArray:(nullable NSString *)string {
    NSObject *object = string.kva_serializedJSONObject;
    return ([object isKindOfClass:NSArray.class] ? (NSArray *) object : nil);
}

// Parse a json serialized NSString into an NSDictionary.
+ (nullable NSDictionary *)parseJsonObject:(nullable NSString *)string {
    NSObject *object = string.kva_serializedJSONObject;
    return [object isKindOfClass:NSDictionary.class] ? (NSDictionary *) object : nil;
}

// Builds and sends an event given an event info dictionary.
+ (void)buildAndSendEvent:(nullable NSDictionary *)eventInfo {
    if(eventInfo == nil) {
        return;
    }
    NSString *name = [KochavaTrackerUtil readNSString:eventInfo[@"name"]];
    NSDictionary *data = [KochavaTrackerUtil readNSDictionary:eventInfo[@"data"]];
    NSString *iosAppStoreReceiptBase64String = [KochavaTrackerUtil readNSString:eventInfo[@"iosAppStoreReceiptBase64String"]];
    if (name.length > 0) {
        KVAEvent *event = [KVAEvent customEventWithNameString:name];
        if (data != nil) {
            event.infoDictionary = data;
        }
        if (iosAppStoreReceiptBase64String.length > 0) {
            event.appStoreReceiptBase64EncodedString = iosAppStoreReceiptBase64String;
        }
        [event send];
    } else {
        [KochavaTrackerUtil log:@"Warn: sendEventWithEvent invalid input"];
    }
}

@end

#pragma mark - Methods

@implementation KochavaTrackerPlugin

// void executeAdvancedInstruction(string name, string value)
- (void)executeAdvancedInstruction:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSString *value = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[1]];
    
    [KVATracker.shared executeAdvancedInstructionWithIdentifierString:name valueObject:value];
}

// void setLogLevel(LogLevel logLevel)
- (void)setLogLevel:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *logLevel = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];

    KVALog.shared.level = [KVALogLevel kva_fromObject:logLevel];
}

// void setSleep(bool sleep)
- (void)setSleep:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    BOOL sleep = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:false];

    KVATracker.shared.sleepBool = sleep;
}

// void setAppLimitAdTracking(bool appLimitAdTracking)
- (void)setAppLimitAdTracking:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    BOOL appLimitAdTracking = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:false];
    
    [KVATracker.shared setAppLimitAdTrackingBool:appLimitAdTracking];
}

// void registerCustomDeviceIdentifier(string name, string value)
- (void)registerCustomDeviceIdentifier:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSString *value = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[1]];
    
    if(name.length > 0 && value.length > 0) {
        [KVATracker.shared.customIdentifiers registerWithNameString:name identifierString:value];
    } else {
        [KochavaTrackerUtil log:@"Warn: Invalid Custom Device Identifier"];
    }
}

// void registerIdentityLink(string name, string value)
- (void)registerIdentityLink:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSString *value = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[1]];

    if(name.length > 0 && value.length > 0) {
        [KVATracker.shared.identityLink registerWithNameString:name identifierString:value];
    } else {
        [KochavaTrackerUtil log:@"Warn: Invalid Identity Link"];
    }
}

// void enableAndroidInstantApps(string instantAppGuid)
- (void)enableAndroidInstantApps:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    [KochavaTrackerUtil log:@"enableAndroidInstantApps API is not available on this platform."];
}

// void enableIosAppClips(string identifier)
- (void)enableIosAppClips:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *identifier = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    
    KVAAppGroups.shared.deviceAppGroupIdentifierString = identifier;
}

// void enableIosAtt()
- (void)enableIosAtt:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    KVATracker.shared.appTrackingTransparency.enabledBool = true;
}

// void setIosAttAuthorizationWaitTime(double waitTime)
- (void)setIosAttAuthorizationWaitTime:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    double waitTime = [KochavaTrackerUtil convertNumberToDouble:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:30.0];
    
    KVATracker.shared.appTrackingTransparency.authorizationStatusWaitTimeInterval = waitTime;
}

// void setIosAttAuthorizationAutoRequest(bool autoRequest)
- (void)setIosAttAuthorizationAutoRequest:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    BOOL autoRequest = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:true];
    
    KVATracker.shared.appTrackingTransparency.autoRequestTrackingAuthorizationBool = autoRequest;
}

// void registerPrivacyProfile(string name, string[] keys)
- (void)registerPrivacyProfile:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSArray *keys = [KochavaTrackerUtil readNSArray:invokedUrlCommand.arguments[1]];
    
    [KVAPrivacyProfile registerWithNameString:name payloadKeyStringArray:keys];
}

// void setPrivacyProfileEnabled(string name, bool enabled)
- (void)setPrivacyProfileEnabled:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    BOOL enabled = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[1]] defaultValue:true];
    
    [KVATracker.shared.privacy setEnabledBoolForProfileNameString:name enabledBool:enabled];
}

// bool getStarted()
- (void)getStarted:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *started = KVATracker.shared.startedBool ? @"true" : @"false";
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:started];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
}

// void start(string androidAppGuid, string iosAppGuid, string partnerName)
- (void)start:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    // arg 0 is androidAppGuid
    NSString *iosAppGuid = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[1]];
    NSString *partnerName = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[2]];

    // Disable osLog and force the use of NSLog.
    KVALog.shared.osLogEnabledBool = false;

    // Register the AdNetwork module. This may eventually not be required but for now include it prior to starting.
    [KVAAdNetworkProduct.shared register];

    if (iosAppGuid.length > 0) {
        [KVATracker.shared startWithAppGUIDString:iosAppGuid];
    } else if (partnerName.length > 0) {
        [KVATracker.shared startWithPartnerNameString:partnerName];
    } else {
        [KochavaTrackerUtil log:@"No App Guid or Partner Name was registered."];
    }
}

// shutdown(bool deleteData)
- (void)shutdown:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    BOOL deleteData = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:true];
    
    [KVATrackerProduct.shared shutdownWithDeleteLocalDataBool:deleteData completionClosure:nil];
}

// string getDeviceId()
- (void)getDeviceId:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *deviceId = @"";
    if(KVATracker.shared.startedBool) {
        deviceId = KVATracker.shared.deviceIdString ?: @"";
    }
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:deviceId];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
}

// InstallAttribution getInstallAttribution()
- (void)getInstallAttribution:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSDictionary *attributionDictionary = [KochavaTrackerUtil convertInstallAttributionToDictionary:KVATracker.shared.attribution.result];
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:attributionDictionary];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
}

//void retrieveInstallAttribution(Callback<InstallAttribution> callback)
- (void)retrieveInstallAttribution:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    [KVATracker.shared.attribution retrieveResultWithCompletionHandler:^(KVAAttributionResult * attribution) {
        NSDictionary *attributionDictionary = [KochavaTrackerUtil convertInstallAttributionToDictionary:attribution];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:attributionDictionary];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
    }];
}

// void processDeeplink(string path, Callback<Deeplink> callback)
- (void)processDeeplink:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSURL *path = [NSURL URLWithString:[KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]]];
    
    [KVADeeplink processWithURL:path completionHandler:^(KVADeeplink * _Nonnull deeplink) {
        NSDictionary *deeplinkDictionary = [KochavaTrackerUtil convertDeeplinkToDictionary:deeplink];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:deeplinkDictionary];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
    }];
}

// void processDeeplinkWithOverrideTimeout(string path, double timeout, Callback<Deeplink> callback)
- (void)processDeeplinkWithOverrideTimeout:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSURL *path = [NSURL URLWithString:[KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]]];
    double timeout = [KochavaTrackerUtil convertNumberToDouble:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[1]] defaultValue:10];
    
    [KVADeeplink processWithURL:path timeoutTimeInterval:timeout completionHandler:^(KVADeeplink * _Nonnull deeplink) {
        NSDictionary *deeplinkDictionary = [KochavaTrackerUtil convertDeeplinkToDictionary:deeplink];
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:deeplinkDictionary];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:invokedUrlCommand.callbackId];
    }];
}

// void registerPushToken(string token)
- (void)registerPushToken:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *token = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    [KVAPushNotificationsToken registerWithDataHexString:token];
}

// void setPushEnabled(bool enabled)
- (void)setPushEnabled:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    BOOL enabled = [KochavaTrackerUtil convertNumberToBool:[KochavaTrackerUtil readNSNumber:invokedUrlCommand.arguments[0]] defaultValue:false];
    KVATracker.shared.pushNotifications.enabledBool = enabled;
}

// void sendEvent(string name)
- (void)sendEvent:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    
    if(name.length > 0) {
        [KVAEvent sendCustomWithNameString:name];
    } else {
        [KochavaTrackerUtil log:@"Warn: sendEvent invalid input"];
    }
}

// void sendEventWithString(string name, string data)
- (void)sendEventWithString:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSString *data = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[1]];
    
    if(name.length > 0) {
        [KVAEvent sendCustomWithNameString:name infoString:data];
    } else {
        [KochavaTrackerUtil log:@"Warn: sendEventWithString invalid input"];
    }
}

// void sendEventWithDictionary(string name, object data)
- (void)sendEventWithDictionary:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSString *name = [KochavaTrackerUtil readNSString:invokedUrlCommand.arguments[0]];
    NSDictionary *data = [KochavaTrackerUtil readNSDictionary:invokedUrlCommand.arguments[1]];
    
    if(name.length > 0) {
        [KVAEvent sendCustomWithNameString:name infoDictionary:data];
    } else {
        [KochavaTrackerUtil log:@"Warn: sendEventWithDictionary invalid input"];
    }
}

// void sendEventWithEvent(Event event)
- (void)sendEventWithEvent:(nonnull CDVInvokedUrlCommand *)invokedUrlCommand {
    NSDictionary *eventInfo = [KochavaTrackerUtil readNSDictionary:invokedUrlCommand.arguments[0]];
    
    [KochavaTrackerUtil buildAndSendEvent:eventInfo];
}

@end
